File: lab3 C++17/src/marker_array.cpp
// marker_array.cpp
//
// lab3 C++17
//
// created by Anhelina Mankewich, 2 course, group 12

#include "marker_array.h"
#include <algorithm>

MarkerArray::MarkerArray(std::size_t size) : data(size, 0) {}

bool MarkerArray::try_mark(std::size_t index, int marker_id) {
    std::lock_guard<std::mutex> lock(mtx);
    if (index >= data.size()) return false;
    if (data[index] == 0) {
        data[index] = marker_id;
        return true;
    }
    return false;
}

void MarkerArray::clear_marks(int marker_id) {
    std::lock_guard<std::mutex> lock(mtx);
    for (auto& value : data) {
        if (value == marker_id) value = 0;
    }
}

std::vector<int> MarkerArray::copy() const {
    std::lock_guard<std::mutex> lock(mtx);
    return data;
}

std::size_t MarkerArray::size() const noexcept {
    return data.size();
}

std::size_t MarkerArray::count_marked_by(int marker_id) const {
    std::lock_guard<std::mutex> lock(mtx);
    return std::count(data.begin(), data.end(), marker_id);
}

int MarkerArray::get(std::size_t index) const {
    std::lock_guard<std::mutex> lock(mtx);
    if (index >= data.size()) return -1;
    return data[index];
}

File: lab3 Java 21/src/MarkerArray.java
// MarkerArray.java
//
// lab3 Java 21
//
// created by Anhelina Mankewich, 2 course, group 12

package src;

import java.util.concurrent.locks.ReentrantLock;

public class MarkerArray {
    private final int[] data;
    private final ReentrantLock lock;

    public MarkerArray(int size) {
        if (size <= 0) {
            throw new IllegalArgumentException("Size must be positive");
        }
        this.data = new int[size];
        this.lock = new ReentrantLock();
    }

    public boolean tryMark(int index, int markerId) {
        lock.lock();
        try {
            if (index < 0 || index >= data.length) {
                return false;
            }
            if (data[index] == 0) {
                data[index] = markerId;
                return true;
            }
            return false;
        } finally {
            lock.unlock();
        }
    }

    public void clearMarks(int markerId) {
        lock.lock();
        try {
            for (int i = 0; i < data.length; i++) {
                if (data[i] == markerId) {
                    data[i] = 0;
                }
            }
        } finally {
            lock.unlock();
        }
    }

    public int[] copy() {
        lock.lock();
        try {
            int[] copy = new int[data.length];
            System.arraycopy(data, 0, copy, 0, data.length);
            return copy;
        } finally {
            lock.unlock();
        }
    }

    public int size() {
        return data.length;
    }

    public int countMarkedBy(int markerId) {
        lock.lock();
        try {
            int count = 0;
            for (int value : data) {
                if (value == markerId) {
                    count++;
                }
            }
            return count;
        } finally {
            lock.unlock();
        }
    }

    public int get(int index) {
        lock.lock();
        try {
            if (index < 0 || index >= data.length) {
                return -1;
            }
            return data[index];
        } finally {
            lock.unlock();
        }
    }
}